File: scripts/install-skill.py
#!/usr/bin/env python3
"""Install one repository skill into Codex, Pi Agent, or a custom directory."""

from __future__ import annotations

import argparse
import sys

from _skill_tools import SkillToolError, install_skill, resolve_source, resolve_target


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("skill", help="skill name or path beneath this repository's skills/")
    target = parser.add_mutually_exclusive_group()
    target.add_argument("--agent", choices=("codex", "pi"), help="installation profile (default: codex)")
    target.add_argument("--target-dir", help="explicit skills directory; overrides agent profiles")
    parser.add_argument(
        "--mode", choices=("link", "copy"), default="link", help="installation mode (default: link)"
    )
    parser.add_argument(
        "--replace",
        action="store_true",
        help="replace an occupied destination after moving it to a timestamped backup",
    )
    parser.add_argument("--dry-run", action="store_true", help="show the resolved operation without writing")
    return parser


def main(argv: list[str] | None = None) -> int:
    args = build_parser().parse_args(argv)
    try:
        source, metadata = resolve_source(args.skill)
        target_root = resolve_target(args.agent, args.target_dir)
        result = install_skill(
            source,
            metadata,
            target_root,
            mode=args.mode,
            replace=args.replace,
            dry_run=args.dry_run,
        )
    except (OSError, SkillToolError) as exc:
        print(f"error: {exc}", file=sys.stderr)
        return 1

    print(f"{result.action}: {metadata.name}")
    print(f"source: {result.source}")
    print(f"destination: {result.destination}")
    if result.backup:
        print(f"backup: {result.backup}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/install-skill.ts
import { argumentsFor, fail, targetConflict } from "./arguments.ts";
import { installSkill, resolveSource, resolveTarget } from "./skill-tools.ts";
try {
  const { values, positionals } = argumentsFor({
    allowPositionals: true,
    options: {
      agent: { type: "string" },
      "target-dir": { type: "string" },
      mode: { type: "string", default: "link" },
      replace: { type: "boolean" },
      "dry-run": { type: "boolean" },
      help: { type: "boolean", short: "h" },
    },
  });
  if (values.help)
    console.log(
      "Usage: node --experimental-strip-types scripts/install-skill.ts SKILL [--agent codex|pi | --target-dir DIRECTORY] [--mode link|copy] [--replace] [--dry-run]",
    );
  else {
    targetConflict(values);
    if (positionals.length !== 1)
      throw new TypeError("Exactly one skill name or path is required");
    if (values.mode !== "link" && values.mode !== "copy")
      throw new TypeError("--mode must be link or copy");
    const [source, metadata] = resolveSource(positionals[0]);
    const result = installSkill(
      source,
      metadata,
      resolveTarget(values.agent, values["target-dir"]),
      { mode: values.mode, replace: values.replace, dryRun: values["dry-run"] },
    );
    console.log(
      `${result.action}: ${metadata.name}\nsource: ${result.source}\ndestination: ${result.destination}`,
    );
    if (result.backup) console.log(`backup: ${result.backup}`);
  }
} catch (error) {
  fail(error);
}
